src/index.js

```diff
@@ -1,98 +1,140 @@
     !Array.isArray(data.x) ||
     data.x.length < 2 ||
     !Array.isArray(data.y) ||
     data.y.length < 2
   ) {
+    throw new Error('The data must have more than 2 points');
+  } else if (data.x.length !== data.y.length) {
     throw new Error(
-      'The data parameter elements must be an array with more than 2 points'
+      'The data object must have equal number of x and y coordinates'
     );
-  } else if (data.x.length !== data.y.length) {
-    throw new Error('The data parameter elements must have the same size');
   }
 
-  var parameters =
-    initialValues || new Array(parameterizedFunction.length).fill(1);
-  let parLen = parameters.length;
+  let params = initialValues || new Array(paramFunction.length).fill(1);
+  let parLen = params.length;
   maxValues = maxValues || new Array(parLen).fill(Number.MAX_SAFE_INTEGER);
   minValues = minValues || new Array(parLen).fill(Number.MIN_SAFE_INTEGER);
 
-  if (maxValues.length !== minValues.length) {
-    throw new Error('minValues and maxValues should have the same size');
+  if (!Array.isArray(params)) {
+    throw new Error('initialValues must be an array');
   }
 
-  if (!Array.isArray(parameters)) {
-    throw new Error('initialValues must be an array');
+  if (
+    maxValues.length !== minValues.length ||
+    maxValues.length !== params.length
+  ) {
+    throw new Error(
+      'minValues and maxValues must have the same size as the number of parameters'
+    );
   }
 
-  var error = errorCalculation(data, parameters, parameterizedFunction);
+  initializeErrorPropagation(errorPropagation);
+
+  /** @type Array<number> */
+  var residualDifferences = Array(10).fill(NaN);
 
-  var converged = error <= errorTolerance;
+  var residuals = sumOfSquaredResiduals(data, params, paramFunction);
+  var converged = false;
 
   for (
     var iteration = 0;
     iteration < maxIterations && !converged;
     iteration++
   ) {
-    parameters = step(
+    var params2 = step(
       data,
-      parameters,
+      params,
       damping,
       gradientDifference,
-      parameterizedFunction
+      paramFunction
     );
 
     for (let k = 0; k < parLen; k++) {
-      parameters[k] = Math.min(
-        Math.max(minValues[k], parameters[k]),
-        maxValues[k]
-      );
+      params2[k] = Math.min(Math.max(minValues[k], params2[k]), maxValues[k]);
     }
 
-    error = errorCalculation(data, parameters, parameterizedFunction);
-    if (isNaN(error)) break;
-    converged = error <= errorTolerance;
+    var residuals2 = sumOfSquaredResiduals(data, params2, paramFunction);
+
+    if (isNaN(residuals2)) throw new Error('The function evaluates to NaN.');
+
+    if (residuals2 < residuals) {
+      params = params2;
+      residuals = residuals2;
+      damping *= dampingDrop;
+    } else {
+      damping *= dampingBoost;
+    }
+
+    damping = Math.max(minDamping, Math.min(maxDamping, damping));
+
+    residualDifferences.shift();
+    residualDifferences.push(residuals - residuals2);
```

Comment:
I think you either need `Math.abs` here or this needs to only happen in the `residuals2 < residuals` case.

Reply:
`Math.abs` wouldn't make sense at all, because it would return large positive values even if the fit hasn't improved. I'm not sure about pushing to the array only *if the fit has improved* either, because `residualEpsilon` is the measure of “how slow can the convergence get before I give up“. And not counting the cases of “not converged at all“ seems like a hell of an unintuitive decission.

Reply:
However I could add this line so that `residualDifference[i]` is zero for the iterations when the fit hasn't improved:
```javascript
if (residuals2 < residuals) {
  params = params2;
  residuals = residuals2;
  damping *= dampingDrop;
} else {
  damping *= dampingBoost;
  residuals2 = residuals; // <--- here!
}
```

Reply:
Doesn't reordering the subtraction so that the result is positive have the same effect as taking the absolute value of the result?

Maybe I still don't understand what is going on well enough yet; please bear with me. The problem I was having was with `converged = maxEpsilon <= residualEpsilon;` triggering termination very early even though the algorithm hadn't converged (but would if I let it keep going). It was my interpretation that `residualEpsilon` was intended to have the following properties:

* Default value of `1e-6` means that if after taking a step the residuals have only improved by `1e-6` the algorithm should stop since it is no longer making significant progress. 
* It represents an absolute value (i.e. it should be >= 0). In that case it didn't make sense to me to be checking if negative numbers are less than it. 
* Setting to `0` should disable the termination condition

BTW, just to check, we using the term *residual* in the [numerical sense](https://en.wikipedia.org/wiki/Residual_(numerical_analysis)) rather than the [statistical sense](https://en.wikipedia.org/wiki/Errors_and_residuals), right?

Reply:
Ah, I think I understand now: the difference is not "magnitude of change" (absolute) but a signed quantity which indicates "improvement" when positive and "deterioration" when negative. Perhaps the problem I was having was because either 10 iterations was not enough history or I had `residualEpsilon` set too small.

Reply:
Yes, that's exactly what it means! And negative values could probably be clamped to `0`, as the algorithm never accepts parameters that are worse than the previous ones. I should add this to the code.

Too small `residualEpsilon` should almost never be a problem – the algorithm would just keep iterating for a little longer. There are could be problems with too large `residualEpsilon`.

But the worst problems I had when testing the function were at the very beginning of the computation. It takes some time before the algorithm corretly adjusts its `damping` parameter and before that, it's often quite slow and tends to halt too quickly (because of low `residualEpsilon`). This could be fixed with a simple `minIterations` option – why haven't I thought of this before?

Reply:
Feel free to try refactoring similar to [my suggestion above](https://github.com/mljs/levenberg-marquardt/pull/19#discussion_r226833326).